README.md
keythereum
==========

[![Build Status](https://travis-ci.org/ethereumjs/keythereum.svg?branch=master)](https://travis-ci.org/ethereumjs/keythereum)
[![Coverage Status](https://coveralls.io/repos/github/ethereumjs/keythereum/badge.svg?branch=master)](https://coveralls.io/github/ethereumjs/keythereum?branch=master)
[![npm version](https://badge.fury.io/js/keythereum.svg)](http://badge.fury.io/js/keythereum)

Keythereum is a JavaScript tool to generate, import and export Ethereum keys.  This provides a simple way to use the same account locally and in web wallets.  It can be used for verifiable cold storage wallets.

Keythereum uses the same key derivation functions (PBKDF2-SHA256 or scrypt), symmetric ciphers (AES-128-CTR or AES-128-CBC), and message authentication codes as [geth](https://github.com/ethereum/go-ethereum).  You can export your generated key to file, copy it to your data directory's keystore, and immediately start using it in your local Ethereum client.

Note: starting in version 0.5.0, keythereum's `encrypt` and `decrypt` functions both return Buffers instead of strings.  This is a breaking change for anyone using these functions directly!

Installation
------------

    $ npm install keythereum

Usage
-----

To use keythereum in Node.js, just `require` it:
```javascript
var keythereum = require("keythereum");
```
A minified, browserified file `dist/keythereum.min.js` is included for use in the browser.  Including this file simply attaches the `keythereum` object to `window`:
```html
<script src="dist/keythereum.min.js" type="text/javascript"></script>
```

### Key creation

Generate a new random private key (256 bit), as well as the salt (256 bit) used by the key derivation function, and the initialization vector (128 bit) used to AES-128-CTR encrypt the key.  `create` is asynchronous if it is passed a callback function, and synchronous otherwise.
```javascript
// optional private key and initialization vector sizes in bytes
// (if params is not passed to create, keythereum.constants is used by default)
var params = { keyBytes: 32, ivBytes: 16 };

// synchronous
var dk = keythereum.create(params);
// dk:
{
    privateKey: <Buffer ...>,
    iv: <Buffer ...>,
    salt: <Buffer ...>
}

// asynchronous
keythereum.create(params, function (dk) {
    // do stuff!
});
```

### Key export

You will need to specify a password and (optionally) a key derivation function.  If unspecified, PBKDF2-SHA256 will be used to derive the AES secret key.
```javascript
var password = "wheethereum";
var kdf = "pbkdf2"; // or "scrypt" to use the scrypt kdf
```
The `dump` function is used to export key info to keystore ["secret-storage" format](https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition).  If a callback function is supplied as the sixth parameter to `dump`, it will run asynchronously:
```javascript
// if options is not passed to dump, it will use the values in
// keythereum.constants by default
var options = {
    kdf: "pbkdf2",
    cipher: "aes-128-ctr",
    kdfparams: {
        c: 262144,
        dklen: 32,
        prf: "hmac-sha256"
    }
};

// synchronous
var keyObject = keythereum.dump(password, dk.privateKey, dk.salt, dk.iv, options);
// keyObject:
{
    address: '008aeeda4d805471df9b2a5b0f38a0c3bcba786b',
    Crypto: {
        cipher: 'aes-128-ctr',
        ciphertext: '5318b4d5bcd28de64ee5559e671353e16f075ecae9f99c7a79a38af5f869aa46',
        cipherparams: {
            iv: '6087dab2f9fdbbfaddc31a909735c1e6'
        },
        mac: '517ead924a9d0dc3124507e3393d175ce3ff7c1e96529c6c555ce9e51205e9b2',
        kdf: 'pbkdf2',
        kdfparams: {
            c: 262144,
            dklen: 32,
            prf: 'hmac-sha256',
            salt: 'ae3cd4e7013836a3df6bd7241b12db061dbe2c6785853cce422d148a624ce0bd'
        }
    },
    id: 'e13b209c-3b2f-4327-bab0-3bef2e51630d',
    version: 3
}

// asynchronous
keythereum.dump(password, dk.privateKey, dk.salt, dk.iv, options, function (keyObject) {
    // do stuff!
});
```
`dump` creates an object and not a JSON string.  In Node, the `exportToFile` method provides an easy way to export this formatted key object to file.  It creates a JSON file in the `keystore` sub-directory, and uses geth's current file-naming convention (ISO timestamp concatenated with the key's derived Ethereum address).
```
keythereum.exportToFile(keyObject);
```
After successful key export, you will see a message like:
```
Saved to file:
keystore/UTC--2015-08-11T06:13:53.359Z--008aeeda4d805471df9b2a5b0f38a0c3bcba786b

To use with geth, copy this file to your Ethereum keystore folder
(usually ~/.ethereum/keystore).
```

### Key import

Importing a key from geth's keystore can only be done on Node.  The JSON file is parsed into an object with the same structure as `keyObject` above.
```javascript
// specify a data directory (optional; defaults to ~/.ethereum)
var datadir = "/home/jack/.ethereum-test";

// synchronous
var keyObject = keythereum.importFromFile(address, datadir);

// asynchronous
keythereum.importFromFile(address, datadir, function (keyObject) {
    // do stuff
});
```
This has been tested with version 3 and version 1, but not version 2, keys.  (Please send me a version 2 keystore file if you have one, so I can test it!)

To recover the plaintext private key from the key object, use `keythereum.recover`.  The private key is returned as a Buffer.
```javascript
// synchronous
var privateKey = keythereum.recover(password, keyObject);
// privateKey:
<Buffer ...>

// asynchronous
keythereum.recover(password, keyObject, function (privateKey) {
    // do stuff
});
```

### Hashing rounds

By default, keythereum uses 65536 hashing rounds in its key derivation functions, compared to the 262144 geth uses by default.  (Keythereum's JSON output files are still compatible with geth, however, since they tell geth how many rounds to use.)  These values are user-editable: `keythereum.constants.pbkdf2.c` is the number of rounds for PBKDF2, and `keythereum.constants.scrypt.n` is the number of rounds for scrypt.

Tests
-----

Unit tests are in the `test` directory, and can be run with mocha:

    $ npm test

`test/geth.js` is an integration test, which is run (along with `test/keys.js`) using:

    $ npm run geth

`geth.js` generates 1000 random private keys, encrypts each key using a randomly-generated passphrase, dumps the encrypted key info to a JSON file, then spawns a geth instance and attempts to unlock each account using its passphrase and JSON file.  The passphrases are between 1 and 100 random bytes.  Each passphrase is tested in both hexadecimal and base-64 encodings, and with PBKDF2-SHA256 and scrypt key derivation functions.

By default, the flags passed to geth are:

    $ geth --etherbase <account> --unlock <account> --nodiscover --networkid "10101" --port 30304 --rpcport 8547 --datadir test/fixtures --password test/fixtures/.password

`test/fixtures/.password` is a file which contains the passphrase.  The `.password` file, as well as the JSON key files generated by `geth.js`, are automatically deleted after the test.

(Note: `geth.js` conducts 4000 tests, each of which can take up to 5 seconds, so running this file can take up to 5.56 hours.)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
(modification: no type change headlines) and this project adheres to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2022-08-03

Maintenance release with various library updates and dependency simplifications.

Note that this version now uses the native JS [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) datatype, which is not supported in some older versions of browsers and dev/build tools!

Following Updates:

1. Replaced `crypto-browserify` with `browserify-aes`, which is a dependency of crypto-browserify; without all the unnecessary modules. We are still using `browserify-aes` because the keythereum's related api methods are synchronous - ethereum-cryptography/aes exposes async-only methods.
2. Replaced `keccak` with `ethereum-cryptography/keccak`
3. Replaced `scrypt-js` with `ethereum-cryptography/scrypt`
4. Replaced `secp256k1` with `ethereum-cryptography/secp256k1-compat`
5. Removed `sjcl` which had pbkdf2 implementation
6. Updated `uuid` from 3.0.0 to 8.3.2
7. The package versions are exact, like before - no version ranges

`wc < dist/keythereum.js` output:
- before `31994  118030  921363`
- after `15243   63140  484003`

Backwards incompatibilities:

- BigInt support is now required
- `keythereum.crypto` is now an emulated object instead of `crypto-browserify` stuff

[2.0.0]: https://github.com/ethereumjs/keythereum/compare/v1.2.0...v2.0.0

## [1.2.0] - 2020-09-29

This is a maintenance release after a longer period with no releases.
See PR [#81](https://github.com/ethereumjs/keythereum/pull/81) for implementation
details.

**Changes**

- `scrypt` to [`scrypt-js`](https://github.com/ricmoo/scrypt-js) for a pure js implementation (simplifies some code)
- `keccak` from `1.4.0` to `3.1.0` for node 12 and n-api support
- `secp256k1` from `3.5.0` to `4.0.2` for node 12 and n-api support
- travis ci node versions from `[4, 5, 6, 7, 8]` to `[8, 10, 12, 13, 14]`
- uglify-js to [`terser`](https://github.com/terser/terser) (build was having some trouble with es6 in node_modules)
- browserify from `16.2.2` to `16.5.2` for misc. bug fixes and upgrades.

[1.2.0]: https://github.com/ethereumjs/keythereum/compare/v1.0.4...v1.2.0

## [1.0.4]

TODO

## Older releases:

- [1.x.x](https://github.com/ethereumjs/keythereum/compare/v1.x.x...v1.y.y) - 20xx-xx-xx
- ...


test/geth.js
/* eslint-env node, mocha */

"use strict";

var fs = require("fs");
var join = require("path").join;
var crypto = require("crypto");
var assert = require("chai").assert;
var geth = require("geth");
var keythereum = require("../");
var checkKeyObj = require("./checkKeyObj");

var NUM_TESTS = 1000;
var TIMEOUT = 10000;
var DATADIR = join(__dirname, "fixtures");

var options = {
  persist: false,
  flags: {
    networkid: "10101",
    port: 30304,
    rpcport: 8547,
    nodiscover: null,
    datadir: DATADIR,
    ipcpath: join(DATADIR, "geth.ipc"),
    password: join(DATADIR, ".password")
  }
};

var pbkdf2 = keythereum.crypto.pbkdf2;
var pbkdf2Sync = keythereum.crypto.pbkdf2Sync;

// geth.debug = true;

function createEthereumKey(passphrase) {
  var dk = keythereum.create();
  var key = keythereum.dump(passphrase, dk.privateKey, dk.salt, dk.iv);
  return JSON.stringify(key);
}

keythereum.constants.quiet = true;

describe("Unlock randomly-generated accounts in geth", function () {
  var password, hashRounds, i;

  var test = function (t) {

    var label = "[" + t.kdf + " | " + t.hashRounds + " rounds] generate key file using password '" + t.password +"'";

    it(label, function (done) {
      var json, keyObject;
      this.timeout(TIMEOUT*2);

      if (t.sjcl) {
        keythereum.crypto.pbkdf2 = undefined;
        keythereum.crypto.pbkdf2Sync = undefined;
      } else {
        keythereum.crypto.pbkdf2 = pbkdf2;
        keythereum.crypto.pbkdf2Sync = pbkdf2Sync;
      }

      json = createEthereumKey(t.password);
      assert.isNotNull(json);

      keyObject = JSON.parse(json);
      assert.isObject(keyObject);
      checkKeyObj.structure(keythereum, keyObject);

      keythereum.exportToFile(keyObject, join(DATADIR, "keystore"), function (keypath) {
        fs.writeFile(options.flags.password, t.password, function (ex) {
          var fail;
          if (ex) return done(ex);
          options.flags.unlock = keyObject.address;
          options.flags.etherbase = keyObject.address;
          geth.start(options, {
            stderr: function (data) {
              if (geth.debug) process.stdout.write(data);
              if (data.toString().indexOf("16MB") > -1) {
                geth.trigger(null, geth.proc);
              }
            },
            close: function () {
              fs.unlink(options.flags.password, function (exc) {
                if (exc) return done(exc);
                fs.unlink(keypath, function (exc) {
                  if (exc) return done(exc);
                  done(fail);
                });
              });
            }
          }, function (err, spawned) {
            if (err) return done(err);
            if (!spawned) return done(new Error("where's the geth?"));
            geth.stdout("data", function (data) {
              var unlocked = "Account '" + keyObject.address+
                "' (" + keyObject.address + ") unlocked.";
              if (data.toString().indexOf(unlocked) > -1) {
                geth.stop();
              }
            });
            geth.stderr("data", function (data) {
              if (data.toString().indexOf("Fatal") > -1) {
                fail = new Error(data);
                geth.stop();
              }
            });
          });
        });
      });
    });
  };

  for (i = 0; i < NUM_TESTS; ++i) {

    password = crypto.randomBytes(Math.ceil(Math.random()*100));
    hashRounds = Math.ceil(Math.random() * 300000);

    keythereum.constants.pbkdf2.c = hashRounds;
    keythereum.constants.scrypt.n = hashRounds;

    test({
      sjcl: false,
      password: password.toString("hex"),
      hashRounds: hashRounds,
      kdf: "pbkdf2"
    });
    test({
      sjcl: true,
      password: password.toString("hex"),
      hashRounds: hashRounds,
      kdf: "pbkdf2"
    });
    test({
      password: password.toString("base64"),
      hashRounds: hashRounds,
      kdf: "scrypt"
    });

    test({
      sjcl: false,
      password: password.toString("hex"),
      hashRounds: hashRounds,
      kdf: "pbkdf2"
    });
    test({
      sjcl: true,
      password: password.toString("hex"),
      hashRounds: hashRounds,
      kdf: "pbkdf2"
    });
    test({
      password: password.toString("base64"),
      hashRounds: hashRounds,
      kdf: "scrypt"
    });
  }

});


test/browser/1.js
(window.webpackJsonp=window.webpackJsonp||[]).push([[1],{284:function(n,e,t){"use strict";t.r(e),t.d(e,"scrypt",function(){return d}),t.d(e,"__wbindgen_throw",function(){return a});var r=t(285);let u=new TextEncoder("utf-8"),f=null;function o(){return null!==f&&f.buffer===r.d.buffer||(f=new Uint8Array(r.d.buffer)),f}function c(n){const e=u.encode(n),t=r.c(e.length);return o().set(e,t),[t,e.length]}let l=new TextDecoder("utf-8");function i(n,e){return l.decode(o().subarray(n,n+e))}let s=null;let w=null;function d(n,e,t,u,f,o){const[l,d]=c(n),[a,b]=c(e),p=(null===s&&(s=r.b()),s);try{r.e(p,l,d,a,b,t,u,f,o);const n=(null!==w&&w.buffer===r.d.buffer||(w=new Uint32Array(r.d.buffer)),w),e=n[p/4],c=n[p/4+1],s=i(e,c).slice();return r.a(e,1*c),s}finally{r.a(l,1*d),r.a(a,1*b)}}function a(n,e){throw new Error(i(n,e))}},285:function(n,e,t){"use strict";var r=t.w[n.i];n.exports=r;t(284);r.f()}}]);

gulpfile.js
"use strict";

var path = require("path");
var cp = require("child_process");
var gulp = require("gulp");
var del = require("del");
var keythereum = require("./");

gulp.task("clean", function (callback) {
  del([path.join("dist", "*.js")], callback);
});

gulp.task("lint", function (callback) {
  cp.exec("npm run lint", function (err, stdout) {
    if (err) if (stdout) process.stdout.write(stdout);
    callback(err);
  });
});

gulp.task("build", function (callback) {
  del([path.join("dist", "*.js")], function (err) {
    if (err) throw err;
    cp.exec("./node_modules/browserify/bin/cmd.js ./exports.js | ./node_modules/uglify-js/bin/uglifyjs > ./dist/keythereum.min.js", function (err, stdout) {
      if (err) throw err;
      if (stdout) process.stdout.write(stdout);
      cp.exec("./node_modules/browserify/bin/cmd.js ./exports.js > ./dist/keythereum.js", function (err, stdout) {
        if (err) throw err;
        if (stdout) process.stdout.write(stdout);
        callback();
      });
    });
  });
});

gulp.task("keygen", function (callback) {
  var dk = keythereum.create();
  var keyobj = keythereum.dump("testpass", dk.privateKey, dk.salt, dk.iv, null);
  keythereum.exportToFile(keyobj, null, function () {
    callback();
  });
});

gulp.task("default", ["lint", "build"]);


index.js
/**
 * Create, import, and export ethereum keys.
 * @author Jack Peterson (jack@tinybike.net)
 */

"use strict";

var isBrowser = typeof process === "undefined" || !process.nextTick || Boolean(process.browser);

var aes = require("browserify-aes");
var scrypt = require("ethereum-cryptography/scrypt");
var ecpbkdf2 = require("ethereum-cryptography/pbkdf2");
var _keccak256 = require("ethereum-cryptography/keccak").keccak256;
var random = require("ethereum-cryptography/random");
var secp256k1 = require("ethereum-cryptography/secp256k1-compat");
var uuid = require("uuid");

function isFunction(f) {
  return typeof f === "function";
}

function keccak256(buffer) {
  return Buffer.from(_keccak256(buffer));
}

module.exports = {

  version: "1.1.0",

  browser: isBrowser,

  scrypt: null,

  crypto: {
    pbkdf2: function (password, salt, iters, dklen, prf, cb) {
      setTimeout(function () {
        ecpbkdf2.pbkdf2(password, salt, iters, dklen, prf).then(function (res) {
          cb(Buffer.from(res));
        });
      }, 0);
    },

    pbkdf2Sync: function (password, salt, iters, dklen, prf) {
      return Buffer.from(ecpbkdf2.pbkdf2Sync(password, salt, iters, dklen, prf));
    },

    randomBytes: function (bytes) {
      return random.getRandomBytesSync(bytes);
    }
  },

  constants: {

    // Symmetric cipher for private key encryption
    cipher: "aes-128-ctr",

    // Initialization vector size in bytes
    ivBytes: 16,

    // ECDSA private key size in bytes
    keyBytes: 32,

    // Key derivation function parameters
    pbkdf2: {
      c: 262144,
      dklen: 32,
      hash: "sha256",
      prf: "hmac-sha256"
    },
    scrypt: {
      memory: 280000000,
      dklen: 32,
      n: 262144,
      r: 8,
      p: 1
    }
  },

  /**
   * Check whether a string is valid hex.
   * @param {string} str String to validate.
   * @return {boolean} True if the string is valid hex, false otherwise.
   */
  isHex: function (str) {
    if (str.length % 2 === 0 && str.match(/^[0-9a-f]+$/i)) return true;
    return false;
  },

  /**
   * Check whether a string is valid base-64.
   * @param {string} str String to validate.
   * @return {boolean} True if the string is valid base-64, false otherwise.
   */
  isBase64: function (str) {
    var index;
    if (str.length % 4 > 0 || str.match(/[^0-9a-z+\/=]/i)) return false;
    index = str.indexOf("=");
    if (index === -1 || str.slice(index).match(/={1,2}/)) return true;
    return false;
  },

  /**
   * Convert a string to a Buffer.  If encoding is not specified, hex-encoding
   * will be used if the input is valid hex.  If the input is valid base64 but
   * not valid hex, base64 will be used.  Otherwise, utf8 will be used.
   * @param {string} str String to be converted.
   * @param {string=} enc Encoding of the input string (optional).
   * @return {Buffer} Buffer (bytearray) containing the input data.
   */
  str2buf: function (str, enc) {
    if (!str || str.constructor !== String) return str;
    if (!enc && this.isHex(str)) enc = "hex";
    if (!enc && this.isBase64(str)) enc = "base64";
    return Buffer.from(str, enc);
  },

  /**
   * Check if the selected cipher is available.
   * @param {string} cipher Encryption algorithm.
   * @return {boolean} If available true, otherwise false.
   */
  isCipherAvailable: function (cipher) {
    return aes.getCiphers().some(function (name) { return name === cipher; });
  },

  /**
   * Symmetric private key encryption using secret (derived) key.
   * @param {Buffer|string} plaintext Data to be encrypted.
   * @param {Buffer|string} key Secret key.
   * @param {Buffer|string} iv Initialization vector.
   * @param {string=} algo Encryption algorithm (default: constants.cipher).
   * @return {Buffer} Encrypted data.
   */
  encrypt: function (plaintext, key, iv, algo) {
    var cipher, ciphertext;
    algo = algo || this.constants.cipher;
    if (!this.isCipherAvailable(algo)) throw new Error(algo + " is not available");
    cipher = aes.createCipheriv(algo, this.str2buf(key), this.str2buf(iv));
    ciphertext = cipher.update(this.str2buf(plaintext));
    return Buffer.concat([ciphertext, cipher.final()]);
  },

  /**
   * Symmetric private key decryption using secret (derived) key.
   * @param {Buffer|string} ciphertext Data to be decrypted.
   * @param {Buffer|string} key Secret key.
   * @param {Buffer|string} iv Initialization vector.
   * @param {string=} algo Encryption algorithm (default: constants.cipher).
   * @return {Buffer} Decrypted data.
   */
  decrypt: function (ciphertext, key, iv, algo) {
    var decipher, plaintext;
    algo = algo || this.constants.cipher;
    if (!this.isCipherAvailable(algo)) throw new Error(algo + " is not available");
    decipher = aes.createDecipheriv(algo, this.str2buf(key), this.str2buf(iv));
    plaintext = decipher.update(this.str2buf(ciphertext));
    return Buffer.concat([plaintext, decipher.final()]);
  },

  /**
   * Derive Ethereum address from private key.
   * @param {Buffer|string} privateKey ECDSA private key.
   * @return {string} Hex-encoded Ethereum address.
   */
  privateKeyToAddress: function (privateKey) {
    var privateKeyBuffer, publicKey;
    privateKeyBuffer = this.str2buf(privateKey);
    if (privateKeyBuffer.length < 32) {
      privateKeyBuffer = Buffer.concat([
        Buffer.alloc(32 - privateKeyBuffer.length, 0),
        privateKeyBuffer
      ]);
    }
    publicKey = Buffer.from(
      secp256k1.publicKeyCreate(privateKeyBuffer, false).slice(1)
    );
    return "0x" + keccak256(publicKey).slice(-20).toString("hex");
  },

  /**
   * Calculate message authentication code from secret (derived) key and
   * encrypted text.  The MAC is the keccak-256 hash of the byte array
   * formed by concatenating the second 16 bytes of the derived key with
   * the ciphertext key's contents.
   * @param {Buffer|string} derivedKey Secret key derived from password.
   * @param {Buffer|string} ciphertext Text encrypted with secret key.
   * @return {string} Hex-encoded MAC.
   */
  getMAC: function (derivedKey, ciphertext) {
    if (derivedKey !== undefined && derivedKey !== null && ciphertext !== undefined && ciphertext !== null) {
      return keccak256(Buffer.concat([
        this.str2buf(derivedKey).slice(16, 32),
        this.str2buf(ciphertext)
      ])).toString("hex");
    }
  },

  /**
   * Used internally.
   */
  deriveKeyUsingScrypt: function (password, salt, options, cb) {
    var n = options.kdfparams.n || this.constants.scrypt.n;
    var r = options.kdfparams.r || this.constants.scrypt.r;
    var p = options.kdfparams.p || this.constants.scrypt.p;
    var dklen = options.kdfparams.dklen || this.constants.scrypt.dklen;
    if (isFunction(cb)) {
      scrypt
        .scrypt(password, salt, n, p, r, dklen)
        .then(function (key) {
          cb(Buffer.from(key));
        })
        .catch(cb);
    } else {
      return Buffer.from(scrypt.scryptSync(password, salt, n, p, r, dklen));
    }
  },

  /**
   * Derive secret key from password with key dervation function.
   * @param {string|Buffer} password User-supplied password.
   * @param {string|Buffer} salt Randomly generated salt.
   * @param {Object=} options Encryption parameters.
   * @param {string=} options.kdf Key derivation function (default: pbkdf2).
   * @param {string=} options.cipher Symmetric cipher (default: constants.cipher).
   * @param {Object=} options.kdfparams KDF parameters (default: constants.<kdf>).
   * @param {function=} cb Callback function (optional).
   * @return {Buffer} Secret key derived from password.
   */
  deriveKey: function (password, salt, options, cb) {
    var prf, iters, dklen;
    if (typeof password === "undefined" || password === null || !salt) {
      throw new Error("Must provide password and salt to derive a key");
    }
    options = options || {};
    options.kdfparams = options.kdfparams || {};

    // convert strings to buffers
    password = this.str2buf(password, "utf8");
    salt = this.str2buf(salt);

    // use scrypt as key derivation function
    if (options.kdf === "scrypt") {
      return this.deriveKeyUsingScrypt(password, salt, options, cb);
    }

    // use default key derivation function (PBKDF2)
    prf = options.kdfparams.prf || this.constants.pbkdf2.prf;
    if (prf === "hmac-sha256") prf = "sha256";
    iters = options.kdfparams.c || this.constants.pbkdf2.c;
    dklen = options.kdfparams.dklen || this.constants.pbkdf2.dklen;
    if (!isFunction(cb)) {
      return Buffer.from(ecpbkdf2.pbkdf2Sync(password, salt, iters, dklen, prf));
    }
    setTimeout(function () {
      ecpbkdf2.pbkdf2(password, salt, iters, dklen, prf).then(function (res) {
        cb(Buffer.from(res));
      });
    }, 0);
  },

  /**
   * Generate random numbers for private key, initialization vector,
   * and salt (for key derivation).
   * @param {Object=} params Encryption options (defaults: constants).
   * @param {string=} params.keyBytes Private key size in bytes.
   * @param {string=} params.ivBytes Initialization vector size in bytes.
   * @param {function=} cb Callback function (optional).
   * @return {Object<string,Buffer>} Private key, IV and salt.
   */
  create: function (params, cb) {
    var keyBytes, ivBytes, self = this;
    params = params || {};
    keyBytes = params.keyBytes || this.constants.keyBytes;
    ivBytes = params.ivBytes || this.constants.ivBytes;

    function checkBoundsAndCreateObject(randomBytes) {
      var privateKey;
      randomBytes = Buffer.from(randomBytes);
      privateKey = randomBytes.slice(0, keyBytes);
      if (!secp256k1.privateKeyVerify(privateKey)) return self.create(params, cb);
      return {
        privateKey: privateKey,
        iv: randomBytes.slice(keyBytes, keyBytes + ivBytes),
        salt: randomBytes.slice(keyBytes + ivBytes)
      };
    }

    // synchronous key generation if callback not provided
    if (!isFunction(cb)) {
      return checkBoundsAndCreateObject(random.getRandomBytesSync(keyBytes + ivBytes + keyBytes));
    }

    // asynchronous key generation
    random.getRandomBytes(keyBytes + ivBytes + keyBytes).then(function (randomBytes) {
      cb(checkBoundsAndCreateObject(randomBytes));
    }, function (err) {
      cb(err);
    });
  },

  /**
   * Assemble key data object in secret-storage format.
   * @param {Buffer} derivedKey Password-derived secret key.
   * @param {Buffer} privateKey Private key.
   * @param {Buffer} salt Randomly generated salt.
   * @param {Buffer} iv Initialization vector.
   * @param {Object=} options Encryption parameters.
   * @param {string=} options.kdf Key derivation function (default: pbkdf2).
   * @param {string=} options.cipher Symmetric cipher (default: constants.cipher).
   * @param {Object=} options.kdfparams KDF parameters (default: constants.<kdf>).
   * @return {Object}
   */
  marshal: function (derivedKey, privateKey, salt, iv, options) {
    var ciphertext, keyObject, algo;
    options = options || {};
    options.kdfparams = options.kdfparams || {};
    algo = options.cipher || this.constants.cipher;

    // encrypt using first 16 bytes of derived key
    ciphertext = this.encrypt(privateKey, derivedKey.slice(0, 16), iv, algo).toString("hex");

    keyObject = {
      address: this.privateKeyToAddress(privateKey).slice(2),
      crypto: {
        cipher: options.cipher || this.constants.cipher,
        ciphertext: ciphertext,
        cipherparams: { iv: iv.toString("hex") },
        mac: this.getMAC(derivedKey, ciphertext)
      },
      id: uuid.v4(), // random 128-bit UUID
      version: 3
    };

    if (options.kdf === "scrypt") {
      keyObject.crypto.kdf = "scrypt";
      keyObject.crypto.kdfparams = {
        dklen: options.kdfparams.dklen || this.constants.scrypt.dklen,
        n: options.kdfparams.n || this.constants.scrypt.n,
        r: options.kdfparams.r || this.constants.scrypt.r,
        p: options.kdfparams.p || this.constants.scrypt.p,
        salt: salt.toString("hex")
      };

    } else {
      keyObject.crypto.kdf = "pbkdf2";
      keyObject.crypto.kdfparams = {
        c: options.kdfparams.c || this.constants.pbkdf2.c,
        dklen: options.kdfparams.dklen || this.constants.pbkdf2.dklen,
        prf: options.kdfparams.prf || this.constants.pbkdf2.prf,
        salt: salt.toString("hex")
      };
    }

    return keyObject;
  },

  /**
   * Export private key to keystore secret-storage format.
   * @param {string|Buffer} password User-supplied password.
   * @param {string|Buffer} privateKey Private key.
   * @param {string|Buffer} salt Randomly generated salt.
   * @param {string|Buffer} iv Initialization vector.
   * @param {Object=} options Encryption parameters.
   * @param {string=} options.kdf Key derivation function (default: pbkdf2).
   * @param {string=} options.cipher Symmetric cipher (default: constants.cipher).
   * @param {Object=} options.kdfparams KDF parameters (default: constants.<kdf>).
   * @param {function=} cb Callback function (optional).
   * @return {Object}
   */
  dump: function (password, privateKey, salt, iv, options, cb) {
    options = options || {};
    iv = this.str2buf(iv);
    privateKey = this.str2buf(privateKey);

    // synchronous if no callback provided
    if (!isFunction(cb)) {
      return this.marshal(this.deriveKey(password, salt, options), privateKey, salt, iv, options);
    }

    // asynchronous if callback provided
    this.deriveKey(password, salt, options, function (derivedKey) {
      cb(this.marshal(derivedKey, privateKey, salt, iv, options));
    }.bind(this));
  },

  /**
   * Recover plaintext private key from secret-storage key object.
   * @param {string|Buffer} password User-supplied password.
   * @param {Object} keyObject Keystore object.
   * @param {function=} cb Callback function (optional).
   * @return {Buffer} Plaintext private key.
   */
  recover: function (password, keyObject, cb) {
    var keyObjectCrypto, iv, salt, ciphertext, algo, self = this;
    keyObjectCrypto = keyObject.Crypto || keyObject.crypto;

    // verify that message authentication codes match, then decrypt
    function verifyAndDecrypt(derivedKey, salt, iv, ciphertext, algo) {
      var key;
      if (self.getMAC(derivedKey, ciphertext) !== keyObjectCrypto.mac) {
        throw new Error("message authentication code mismatch");
      }
      if (keyObject.version === "1") {
        key = keccak256(derivedKey.slice(0, 16)).slice(0, 16);
      } else {
        key = derivedKey.slice(0, 16);
      }
      return self.decrypt(ciphertext, key, iv, algo);
    }

    iv = this.str2buf(keyObjectCrypto.cipherparams.iv);
    salt = this.str2buf(keyObjectCrypto.kdfparams.salt);
    ciphertext = this.str2buf(keyObjectCrypto.ciphertext);
    algo = keyObjectCrypto.cipher;

    if (keyObjectCrypto.kdf === "pbkdf2" && keyObjectCrypto.kdfparams.prf !== "hmac-sha256") {
      throw new Error("PBKDF2 only supported with HMAC-SHA256");
    }

    // derive secret key from password
    if (!isFunction(cb)) {
      return verifyAndDecrypt(this.deriveKey(password, salt, keyObjectCrypto), salt, iv, ciphertext, algo);
    }
    this.deriveKey(password, salt, keyObjectCrypto, function (derivedKey) {
      try {
        cb(verifyAndDecrypt(derivedKey, salt, iv, ciphertext, algo));
      } catch (exc) {
        cb(exc);
      }
    });
  },

  /**
   * Generate filename for a keystore file.
   * @param {string} address Ethereum address.
   * @return {string} Keystore filename.
   */
  generateKeystoreFilename: function (address) {
    var filename = "UTC--" + new Date().toISOString() + "--" + address;

    // Windows does not permit ":" in filenames, replace all with "-"
    if (process.platform === "win32") filename = filename.split(":").join("-");

    return filename;
  },

  /**
   * Export formatted JSON to keystore file.
   * @param {Object} keyObject Keystore object.
   * @param {string=} keystore Path to keystore folder (default: "keystore").
   * @param {function=} cb Callback function (optional).
   * @return {string} JSON filename (Node.js) or JSON string (browser).
   */
  exportToFile: function (keyObject, keystore, cb) {
    var outfile, outpath, json, fs;
    keystore = keystore || "keystore";
    outfile = this.generateKeystoreFilename(keyObject.address);
    json = JSON.stringify(keyObject);
    if (this.browser) {
      if (!isFunction(cb)) return json;
      return cb(json);
    }
    outpath = require("path").join(keystore, outfile);
    fs = require("fs");
    if (!isFunction(cb)) {
      fs.writeFileSync(outpath, json);
      return outpath;
    }
    fs.writeFile(outpath, json, function (err) {
      if (err) return cb(err);
      cb(outpath);
    });
  },

  /**
   * Import key data object from keystore JSON file.
   * (Note: Node.js only!)
   * @param {string} address Ethereum address to import.
   * @param {string=} datadir Ethereum data directory (default: ~/.ethereum).
   * @param {function=} cb Callback function (optional).
   * @return {Object} Keystore data file's contents.
   */
  importFromFile: function (address, datadir, cb) {
    var keystore, filepath, path, fs;
    if (this.browser) throw new Error("method only available in Node.js");
    path = require("path");
    fs = require("fs");
    address = address.replace("0x", "");
    address = address.toLowerCase();

    function findKeyfile(keystore, address, files) {
      var i, len, filepath = null;
      for (i = 0, len = files.length; i < len; ++i) {
        if (files[i].indexOf(address) > -1) {
          filepath = path.join(keystore, files[i]);
          if (fs.lstatSync(filepath).isDirectory()) {
            filepath = path.join(filepath, files[i]);
          }
          break;
        }
      }
      return filepath;
    }

    datadir = datadir || path.join(process.env.HOME, ".ethereum");
    keystore = path.join(datadir, "keystore");
    if (!isFunction(cb)) {
      filepath = findKeyfile(keystore, address, fs.readdirSync(keystore));
      if (!filepath) {
        throw new Error("could not find key file for address " + address);
      }
      return JSON.parse(fs.readFileSync(filepath));
    }
    fs.readdir(keystore, function (ex, files) {
      var filepath;
      if (ex) return cb(ex);
      filepath = findKeyfile(keystore, address, files);
      if (!filepath) {
        return cb(new Error("could not find key file for address " + address));
      }
      return cb(JSON.parse(fs.readFileSync(filepath)));
    });
  }

};
